### PR TITLE
add NodeKernelDeadlock condition

### PR DESCRIPTION
**Pull request by auhlig**
Adding KernelDeadlock condition as introduced by the node problem detector [here](https://github.com/kubernetes/node-problem-detector/blob/master/config/kernel-monitor.json#L12).

### CODE REVIEW

**Comment by grodrigues3** [comment]
@k8s-bot ok to test

**Comment by auhlig** [comment]
Grrr.. Will fix tomorrow.

**Comment by auhlig** [comment]
Fixed by running the `hack/update-staging-client-go.sh`script, which was somewhat painful on mac. I guess the pending submit queue is not related to this PR, @grodrigues3.

**Review by majewsky — APPROVED** [review]

**Comment by auhlig** [comment]
Could you take a look at this PR @grodrigues3, @errordeveloper?  Would be highly appreciated.

**Comment by andyxning** [comment]
@Random-Liu 

Maybe we need to also add a specification or plan for what node conditions node-problem-detector will add.

File: pkg/api/v1/types.go [hunk 1]
```diff
@@ -3291,6 +3291,8 @@ const (
 	NodeNetworkUnavailable NodeConditionType = "NetworkUnavailable"
 	// NodeInodePressure means the kubelet is under pressure due to insufficient available inodes.
 	NodeInodePressure NodeConditionType = "InodePressure"
+	// NodeKernelDeadlock means the kernel of a node has a temporary or permanent deadlock.
```
Comment (andyxning) [inline pkg/api/v1/types.go]: This should be clearly documented that `NodeKernelDeadlock` is updated from node-problem-detector. In order to get this condition, users should deploy a node-problem-detector. :)

Comment (auhlig) [inline pkg/api/v1/types.go]: Makes sense. Done via 42fcdab.

**Comment by Random-Liu** [comment]
@auhlig Actually it's not settled down yet how we should deal with the extra conditions introduced by other components (e.g. NPD).

For NPD, the conditions introduced are currently purely config-file-based. It doesn't seem to make sense to add it to Kubernetes api, if NPD hasn't enforced it yet.

At least, before adding this to Kubernetes api, NPD should maintain a list of valid conditions, and make sure the conditions in the config file must be valid.

@dchen1107 WDYT?

**Comment by auhlig** [comment]
Thanks for clarifying @Random-Liu.
Since the`KernelDeadlock` node condition was part of the latest stable release I assumed would be valid condition.

**Comment by dchen1107** [comment]
@auhlig Thanks for sending the pr. But I think we are not planning to introduce this as another Node Condition due to the following reasons:

* We are in the process to switching the node conditions to taints. Using node condition or events was a temporary solution, and the decision was made long time ago at: https://github.com/kubernetes/kubernetes/issues/23028#issuecomment-215578728
* NPD is designed and implemented as the daemon collecting and propagating the node problems. It should be easy to extended to integrate with other problem-detector. Today there is only kernel-monitor as the default problem-detector, but the K8S vendor / admin can develop their own plugins, for example, on-prem, bad-hardware-detector. Each problem-detector will introduce their own set of problems. We don't want all those problems are included in the core APIs
* Even for today's kernel-monitor, it is configure-based which is really subject to change based on the kernel version, module and kernel configuration. For example, we can configure kernel-monitor to detect kernel hung task. But if your guest kernel enable kernel.hung_task_panic and pvpanic, there is no way for the kernel-monitor to collect such information. Instead we need to either collect those from either serial console or pvpanic events outside the node. 
* Based on all above reasons, we decided to not provide a common remedy system for K8S stack. Instead the K8S vendor / admin(s) can implement their own remedy system along with re-configuring kernel-monitor, adding new problem-detector plugins, etc. based on their own production environment. 

I am suggesting to close this pr. Thanks! cc/ @kubernetes/sig-node-feature-requests

**Comment by auhlig** [comment]
Makes sense. Thanks for explaining @dchen1107 . Closing.

**Pull request closed without merging**